L1Trigger/L1TTrackMatch/plugins/L1TkElectronTrackProducer.cc

```diff
@@ -0,0 +1,340 @@
+  maxChi2IsoTracks = iConfig.getParameter<double>("maxChi2IsoTracks");
+  minNStubsIsoTracks = iConfig.getParameter<int>("minNStubsIsoTracks");
+  // cut applied on the isolation (if this number is <= 0, no cut is applied)
+  IsoCut = (float)iConfig.getParameter<double>("IsoCut");
+  RelativeIsolation = iConfig.getParameter<bool>("RelativeIsolation");
+
+  // parameters to select tracks to match with L1EG
+  trkQualityChi2 = (float)iConfig.getParameter<double>("TrackChi2");
+  trkQualityPtMin = (float)iConfig.getParameter<double>("TrackMinPt");
+  useTwoStubsPT = iConfig.getParameter<bool>("useTwoStubsPT");
+  dPhiCutoff = iConfig.getParameter<std::vector<double> >("TrackEGammaDeltaPhi");
+  dRCutoff = iConfig.getParameter<std::vector<double> >("TrackEGammaDeltaR");
+  dEtaCutoff = iConfig.getParameter<std::vector<double> >("TrackEGammaDeltaEta");
+  matchType = iConfig.getParameter<std::string>("TrackEGammaMatchType");
+
+  produces<TkElectronCollection>(label);
+}
+
+L1TkElectronTrackProducer::~L1TkElectronTrackProducer() {}
+
+// ------------ method called to produce the data  ------------
+void L1TkElectronTrackProducer::produce(edm::Event& iEvent, const edm::EventSetup& iSetup) {
+  std::unique_ptr<TkElectronCollection> result(new TkElectronCollection);
+
+  // geometry needed to call pTFrom2Stubs
+  edm::ESHandle<TrackerGeometry> geomHandle;
+  iSetup.get<TrackerDigiGeometryRecord>().get("idealForDigi", geomHandle);
+  const TrackerGeometry* tGeom = geomHandle.product();
+
+  // the L1EGamma objects
+  edm::Handle<EGammaBxCollection> eGammaHandle;
+  iEvent.getByToken(egToken, eGammaHandle);
+  EGammaBxCollection eGammaCollection = (*eGammaHandle.product());
+  EGammaBxCollection::const_iterator egIter;
+
+  // the L1Tracks
+  edm::Handle<L1TTTrackCollectionType> L1TTTrackHandle;
+  iEvent.getByToken(trackToken, L1TTTrackHandle);
+  L1TTTrackCollectionType::const_iterator trackIter;
+
+  if (!eGammaHandle.isValid()) {
+    edm::LogError("L1TkElectronTrackProducer") << "\nWarning: L1EmCollection not found in the event. Exit" << std::endl;
+    return;
+  }
+  if (!L1TTTrackHandle.isValid()) {
+    edm::LogError("TkEmProducer") << "\nWarning: L1TTTrackCollectionType not found in the event. Exit." << std::endl;
+    return;
+  }
+
+  int ieg = 0;
+  for (egIter = eGammaCollection.begin(0); egIter != eGammaCollection.end(0); ++egIter) {  // considering BX = only
+    edm::Ref<EGammaBxCollection> EGammaRef(eGammaHandle, ieg);
+    ieg++;
+
+    float e_ele = egIter->energy();
+    float eta_ele = egIter->eta();
+    float et_ele = 0;
+    if (cosh(eta_ele) > 0.0)
+      et_ele = e_ele / cosh(eta_ele);
+    else
+      et_ele = -1.0;
+    if (ETmin > 0.0 && et_ele <= ETmin)
+      continue;
+    // match the L1EG object with a L1Track
+    float drmin = 999;
+    int itr = 0;
+    int itrack = -1;
+    for (trackIter = L1TTTrackHandle->begin(); trackIter != L1TTTrackHandle->end(); ++trackIter) {
```

Comment:
range-based loop

Reply:
leave as is to avoid changing code beloe

Reply:
all this does is change an iterator to a const ref to the underlying object, avoiding temporaries and indirection. changes to the loop code would be minimal.